docs/getting-started/tutorial.md

```diff
@@ -0,0 +1,349 @@
+        }
+      }
+    },
+  }
+...
+```
+
+Some properties as passed within a `ui:options` object within that field, `labels` included. You can find out more about which properties are available and how they get passed to `uiSchema` by reading more in [about the schema and uiSchema objects](../building-a-form/about-the-schema-and-uischema-objects.md).
+
+### Step 10: Add specific data constraints and validation
+
+The final question in our form is for the ZIP code, which is also the most complicated question. While ZIP codes are usually `string` data, they also need to follow specific formats. How do we specify data contraints in our `formConfig`?
+
+The primary way to do so is through the `schema` object. The JSON Schem specification includes multiple properties to add constraints to the data that is accepted for that field: `minLength`, `maxLength`, `format`, and `pattern`, among others.
+
+We're going to be using the `pattern` property to validate our ZIP code input, which takes a regex to match the user entry against.
+
+This is what our `schema` looks like with the ZIP code field added:
+```js
+...
+  schema: {
+    type: 'object',
+    properties: {
+      street: {
+        type: 'string'
+      },
+      city: {
+        type: 'string'
+      },
+      state: {
+        type: 'string',
+        enum: ['AK', 'IL', 'MA', 'TX', 'WY']
+      },
+      zip: {
+        type: 'string',
+        pattern: '^[0-9]{5,9}$'
+      }
+    }
+  },
+...
+```
+
+Wonderful! Now we have a ZIP code field. Let's test entering an invalid ZIP code. Enter an invalid entry and focus away from the ZIP code field:
+![ZIP code pattern error](../images/zip-pattern-error.png)
+
+Uh oh! While the form caught that the entry wasn't correct, and displayed an error to the user, the error isn't particularly readable.
+
+We can add custom error messages to our form for specific cases. Again, since this is related to what is displayed to the user, custom error messages go in the `uiSchema`:
+
+```js
+...
+  uiSchema: {
+    street: {
+      'ui:title': 'Street'
+    },
+    city: {
+      'ui:title': 'City'
+    },
+    state: {
+      'ui:title': 'State',
+      'ui:options': {
+        labels: {
+          AK: 'Alaska',
+          IL: 'Illinois',
+          MA: 'Massachusetts',
+          TX: 'Texas',
+          WY: 'Wyoming'
+        }
+      }
+    },
+    zip: {
+      'ui:title': 'ZIP code',
+      'ui:errorMessages': {
+        'pattern': 'Please enter a valid US ZIP code. It must be between 5 and 9 numbers.'
+      }
+    }
+  }
+...
+```
+
+In order to display a custom error message, we pass the `ui:errorMessages` object to `uiSchema` under the field for `zip`. The `ui:errorMessages` object takes key-value pairs, with the key being the name of the JSON Schema property the entry is in violation of, the value being the message we want displayed in that case. By doing this, we can add much more description error messages to the form to improve the user experience.
```

Comment:
> In order to display a custom error message, we pass the `ui:errorMessages` object to `uiSchema` under the field for `zip`. The `ui:errorMessages` object takes key-value pairs, with the key being the name of the JSON Schema property the entry is in violation of, the value being the message we want displayed in that case. By doing this, we can add much more description error messages to the form to improve the user experience.

> Now we're seeing a much more useful error message:

A little tighter:

"To display a custom error message, you pass the `ui:errorMessages` object to `uiSchema` under the field for `zip`. `ui:errorMessages` takes key-value pairs, where the key is the name of the JSON Schema property that the entry violates, and the value is the user-friendly, descriptive message displayed for the violation.

Now you'll see a much more useful error message:"